### ISSUE TITLE
Symmetry constraints API discussion

### ISSUE THREAD

**Issue by tristanic** [openmm/openmm#3757]
Following on from our offline discussion, opening a new ticket to discuss what an API might look like for implementations of (1) full PME simulations based on a small asymmetric unit (i.e. space group - #1437, point group, helical and wallpaper symmetry) and (2) more general symmetry constraints without periodic boundary conditions (#1966).

I can of course only speak to my use cases for these: 

Case (1) would be used primarily for non-interactive settling/equilibration of a model into its experimentally-derived map(s), ensuring all symmetry interfaces are explicitly modelled. Using strict symmetry via virtual sites rather than explicit copies is important in this case because early-stage models often need extensive rebuilding; with explicit copies it is very difficult to propagate corrections throughout all copies. In the context of ISOLDE such simulations would always take place in implicit solvent - as discussed in #1437, imposing rotational symmetry on solvent molecules seems unwise at best.

In the absence of explicit solvent, periodic boundary conditions would only strictly be needed on (a) all three axes for space group symmetry, (b) two axes for wallpaper symmetry, (c) one axis for helical symmetry, and (c) not at all for point group symmetry. In practice, I'm not sufficiently familiar with PME implementations to know what makes the most sense here.

Case (2) is what I envision being used for most interactive work in ISOLDE. See the below image (I've excluded the associated map for clarity). The "real" asymmetric unit is the lighter-coloured one on left; the darker one is a crystal copy. The visible atoms would be mobile in a simulation, with an extra shell of rigidly fixed atoms around them to keep everything in its wider context. Periodic boundary conditions are not used. Most ISOLDE simulations are constructed in this way (albeit not usually spherical like this): the user-chosen selection (highlighted in green) is extended outwards by three residues along all covalent bonds; this selection is then expanded outwards to include all residues with atoms within 5 Å of any selected atom; and a fixed shell is defined by the result of a second outward expansion. On a good Nvidia GPU, simulation performance begins to limit interactive work at around 1500-2000 amino acid residues. 

At present, the symmetry atoms are not included in simulations but are visualised as "ghosts", leaving it up to the user to add restraints as necessary to keep interfaces under control. This visualisation is currently only provided in space group symmetry settings (i.e. for crystallographic models), but I will be adding support for other symmetry cases.

In terms of API for this case, I would think the sensible approach would be for the caller to be responsible for providing a list of rotation/translation matrices defining the relevant transforms (including the identity transform), along with a {residue: [transform indices]} map defining which transform(s) should be applied to each residue in the simulation (at least, it seems most sensible to me to define the constraints on a residue-by-residue rather than atom-by-atom basis). Possibly a little more tricky, but would be important to include a mechanism for exclusions, to deal with cases like https://github.com/openmm/openmm/issues/1437#issuecomment-581939208 where there is a localised breakdown in symmetry.
  
![model_region](https://user-images.githubusercontent.com/22666848/187418531-a3470d61-21c5-45e1-8955-650f03cb43ef.jpg)

**Comment by peastman** [comment]
Doing this with virtual sites would be straightforward.  We could add a new `SymmetricSite` subclass of `VirtualSite`.  The constructor would take a single particle index to base it on, plus a rotation matrix and translation vector.

> Possibly a little more tricky, but would be important to include a mechanism for exclusions, to deal with cases like https://github.com/openmm/openmm/issues/1437#issuecomment-581939208 where there is a localised breakdown in symmetry.

You suggested handling this by having two versions of the residue, and adding nonbonded exclusions to make sure each one only interacted with the virtual site copies of the other one.  Would that meet your requirements?  An alternative would be to have a single copy, and define the virtual sites in a way so they always adopt the other conformation.  That ought to be possible in principle, but perhaps very complicated in practice!

**Comment by tristanic** [comment]
> You suggested handling this by having two versions of the residue, and adding nonbonded exclusions to make sure each one only interacted with the virtual site copies of the other one. Would that meet your requirements?

Yes, I think that still seems like the most sensible approach.

**Comment by tristanic** [comment]
That sounds like it will work for me!

**Comment by tristanic** [comment]
Just thought it worth sharing this perfect example of a use-case for arbitrary symmetry constraints: https://twitter.com/BJ_Greber/status/1565001897286803458?s=20&t=bHZ24esT7itXflIXVcFZDw. The complexes in question are over 6 MDa, but the actual number of unique species is quite small. Each of the six "arms" is a trimer of hexamers - so 108 copies of the same chain. As you can imagine, finding an error in one copy and having to propagate the fix to the other 107 would be enormously painful... a much nicer way to go would be to start with all copies symmetry-constrained, and only remove the constraints (potentially replacing with symmetry _re_straints) once satisfied that any systematic errors are dealt with.

**Comment by aizvorski** [comment]
Hi guys - I'm very happy to hear this is moving forward!  I'm interested in being able to use especially rotational and helical symmetry - both would be (typically) periodic along one direction only.  Happy to help test this.  What would the proposed SymmetricSite api look like in those cases?  Specified with Rx + ua + vb + wc for each site, or something else?

**Comment by peastman** [comment]
What do `ua`, `vb`, and `wc` represent?

**Comment by aizvorski** [comment]
@peastman Ah, that was quoting from what you mentioned in this thread: https://github.com/openmm/openmm/issues/1437#issuecomment-202616443 

u, v, w are (scalar) coefficients and a, b, c the unit cell vectors

**Comment by peastman** [comment]
That's a bit different from what I suggested above.  That was just `Rx+v` where `v` is a fixed vector.  These seem like two distinct use cases: one that's mainly for non-periodic systems, or for modeling multimers that don't depend on the unit cell dimensions, and the other for constructing periodic copies that do depend on it and should change if the unit cell changes size.  Which one is relevant for you?

**Comment by aizvorski** [comment]
@peastman Hmm, as you said, the fixed vector method is good for symmetric assemblies with a fixed number of units (eg a dimer); the cell-size-dependent method is good for assemblies that are periodic in at least one direction.  I'm most interested in spiral structures at the moment, but I think it's good to have both capabilities since neither one is a perfect substitute for the other.  Would it be hard to write the api so it can be either cell-size dependent or fixed? (either two different classes or a class that has both parameters per site, just zero out the unused one)

P.P.S. Having the cell-size-dependent api opens a straightforward path to read in pdb or mmcif file and create a simulaiton of the protein crystal including the correct crystal packing - I think Tristan would like that :)

**Comment by peastman** [comment]
I'd like to start work on this.  I'm thinking of the following API for the constructor:

```c++
SymmetrySite(int particle, const Vec3& Rx, const Vec3& Ry, const Vec3& Rz, const Vec3& v, bool useBoxVectors)
```

`Rx`, `Ry`, and `Rz` make up the rotation matrix $\mathbf{R}$.  If `useBoxVectors` is false then $\mathbf{v}$ is treated as a simple vector.  The new position is calculated as

$\mathbf{Rx} + \mathbf{v}$

where $\mathbf{x}$ is the current position.  If `useBoxVectors` is true, then $\mathbf{v}$ is instead treated as a set of scale factors for the periodic box vectors.  The new position is

$\mathbf{Rx} + v_x\mathbf{a} + v_y\mathbf{b} + v_z\mathbf{c}$

Does that sound good?

**Comment by peastman** [comment]
In testing this, I discovered something interesting.  If you add a SymmetrySite, then momentum is no longer conserved.  I should have anticipated that, but I didn't think about it.  The rotation matrix rotates the position around the origin.  If you translate the whole system, that means it gets rotated around a different point relative to the system.  It breaks translational invariance.

Angular momentum also isn't conserved, for a similar reason.  The position gets translated by a fixed vector.  If you rotate the system, the translation direction changes relative to the rotated system.  It breaks rotational invariance.

I guess this is all as expected, but is it a problem?  Do we need to reconsider the design in light of it?  Or is this just what we expect, and it's not a problem?

**Comment by peastman** [comment]
Thanks!  Any thoughts about the API?  Should it offer both fixed and box-relative offsets at the same time, or just a choice of one versus the other?

**Comment by tristanic** [comment]
I’m not sure I totally understand the question. The example you gave looks
like the distinction between non-crystallographic and crystallographic
symmetry (multiple copies packing an asymmetric unit, then multiple copies
of *that* packing the unit cell). Personally I try to avoid strict NCS
(it’s rare to find cases where there *aren’t* subtle systematic differences
between copies), but I’m sure it would be useful on occasion. But then
there are cases like helical and icosahedral reconstructions where strict
non-crystallographic symmetry makes perfect sense.

**Comment by aizvorski** [comment]
@peastman Thanks!  This is great, also very excited you're working on it

> Alternatively, should there be two separate offset vectors, one that's a simple offset and one that scales the box vectors?

Ah, I was just trying to puzzle that out myself :)

I can think of a few pretty different use cases, let me try to walk through them:
- Molecular crystals: cif files which can be read by [Gemmi](https://gemmi.readthedocs.io/en/latest/index.html).  The symmetry within a unit cell is described [in this style](https://gemmi.readthedocs.io/en/latest/symmetry.html#operations): `x,x-y,z+1/2` where x,y,z is the position projected on the unit cell vectors (linear combinations of the vectors are possible, so this seems to be a 3x3 matrix) => I think this can be represented as "rotation matrix using coordinates expressed in terms of box vectors", but probably not easily as "rotation in cartesian coordinates"
- Protein crystals (eg from rcsb): when saved as a mmcif file they also have a symmetry operation in the file, but that seems to be always x,y,z (identity), multiple copies of the protein within a cell are represented as explicit duplicated copies (?) => these can probably be loaded and simulated already without using SymmetrySite, but this won't preserve symmetry; there would have to be a step that detects the symmetry and replaces copies with symmetry sites.  If replaced with symmetry sites then again would need "rotation matrix using coordinates expressed in terms of box vectors"
- Proteins during crystallography work: probably again "rotation matrix using coordinates expressed in terms of box vectors", but Tristan would have a better idea about that
- Weird constructs: I'm personally interested in the dynamics of helical structures (eg an infinite alpha helix, or in a more complicated case, microtubules like in [5JCO](https://www.rcsb.org/structure/5JCO)).  It seems in practice this would have an unavoidable limitation where some number of units in the helix have to add up to an integer number of full turns (units/turn is an integer fraction) so the box can repeat by translation along the helical axis; during a simulation the units-per-turn can't change, although the pitch can change.  This would be handled pretty well by "rotation matrix in terms of box vectors" + "translation in terms of box vectors".

It's hard to think of a use case which truly needs a mix of both "symmetry in terms of the box vectors" and "cartesian symmetry" (? still thinking about this)

In short, I think there could be two kinds of SymmetrySites, BoxSymmetrySite and CartesianSymmetrySite (not exactly sure of the best naming, so please suggest better names :)   They're both Rx+v, but R and v are expressed in terms of fractional coordinates projected on the box vectors for BoxSymmetrySite, and plain old cartesian coordinates for CartesianSymmetrySite.  A molecular crystal, protein crystal, and helix would all be able to use BoxSymmetrySite; something like a non-periodic simulation of a dimer with C2 symmetry could be handled with CartesianSymmetrySite, assuming the dimer is centered on the origin.  If we truly need a mix, we could maybe have a CartesianSymmetrySite that depends on a BoxSymmetrySite.

**Comment by peastman** [comment]
I mean, which of the following APIs is better?

```c++
SymmetrySite(int particle, const Vec3& Rx, const Vec3& Ry, const Vec3& Rz, const Vec3& v, bool useBoxVectors)
```

or

```c++
SymmetrySite(int particle, const Vec3& Rx, const Vec3& Ry, const Vec3& Rz, const Vec3& v, const Vec3& boxScale)
```

I lean toward the second one since it's more flexible.

**Comment by tristanic** [comment]
Yep - the second one works for me!

**Comment by peastman** [comment]
> They're both Rx+v, but R and v are expressed in terms of fractional coordinates projected on the box vectors

I don't see how that could work.  Unless your box happens to be a cube, the rotation would distort your molecule.  It would get stretched in the directions where the box is bigger and squashed in the directions where the box is smaller.

**Comment by aizvorski** [comment]
@peastman Hmm, that's true, but this seems to be how cif files think about it.  I think they have a mix of reflections and translations, neither of which would change volume (?).

For example [acetamide in CCSD](https://www.ccdc.cam.ac.uk/structures/search?id=doi:10.5517/cc7kw7w&sid=DataCite) has

```
1 x,y,z
2 1/3+x,2/3+y,2/3+z
3 2/3+x,1/3+y,1/3+z
4 -y,x-y,z
5 -x+y,-x,z
6 1/3-y,2/3+x-y,2/3+z
7 1/3-x+y,2/3-x,2/3+z
8 2/3-y,1/3+x-y,1/3+z
...
```

and indeed for all of these, it seems x,y,z are only ever multiplied by +-1 (or 0); the offsets are integer fractions.

**Comment by tristanic** [comment]
As I understand it, that’s the key constraint that limits the number of
possible space groups: the combination of cell shape and
fractional-coordinates transformation matrix must yield a rigid-body
transform in both fractional and Cartesian coordinates.

**Comment by peastman** [comment]
So rotations in cartesian space and translations in terms of the box vectors would work?

**Comment by tristanic** [comment]
In theory, yes - but I'm not sure if I'd do it that way. The various
crystallographic libraries (Clipper - that I use in ChimeraX; CCTBX; Gemmi;
etc.) all provide APIs for converting transformation matrices from
fractional to Cartesian coordinates - so having the rotations in Cartesian
and translations in box vector coordinates would probably just confuse
things.

**Comment by peastman** [comment]
If you're just concerned about the initial system setup, it's easy to convert either way between Cartesian and fractional.  The question is, how should the virtual site move as the base particle moves and as the box size changes?

**Comment by aizvorski** [comment]
> The question is, how should the virtual site move as the base particle moves and as the box size changes?

@peastman For minimization, it should probably move in fractional coordinates.

For dynamics, if studying something like thermal expansion or compressibility (or trying to equilibrate all the water in a protein crystal)  probably still fractional coordinates?  Maybe for a more realistic simulation those might be better done using a supercell full of individual copies without symmetry, because the motions of next neighbors aren't that correlated, but if it is desirable to preserve some of the symmetry exactly, then still fractional.

**Comment by peastman** [comment]
> For dynamics, if studying something like thermal expansion or compressibility (or trying to equilibrate all the water in a protein crystal) probably still fractional coordinates?

Can you explain why?  Here's how I'm thinking of it.  You run a simulation of multiple molecules that are rotated and translated copies of a single molecule.  Assume your periodic box is not a cube, and also that its size/shape may change during the simulation according to a barostat.  The one primary molecule will be realistic.  Its bonds will have the correct lengths and atoms will be the correct distances apart.  But if you perform rotations in fractional coordinates, then all the other copies will be distorted.  Bonds will have the wrong lengths.  Some atoms will be too close together and others too far apart.

Of course, if you only allow rotations of 180 degrees you'll be ok.  But any other rotation will distort the molecule.  Think of a hexamer composed of six copies rotated by pi/3.  And the cases that still work with fractional coordinates are simply the cases where it doesn't matter whether you use fractional or Cartesian coordinates.  They both give the same result.  But in any case where they give different results, fractional coordinates will produce a distorted molecule.

**Comment by aizvorski** [comment]
@peastman Let me first say thank you for a very good clarifying discussion, this has been great for thinking through how to use this kind of feature as well as about the api.

For something like a molecular crystal at/near 0K with a barostat, let's say with a triclinic box; all the positions are described in terms of the primary molecule using only 180 deg rotations (although I think it's only 180 around the box vectors - not sure if it can be 180 around arbitrary linear combination of box vectors?).  We want everything to stay in the same fractional positions as the barostat adjusts the box vectors and possibly the angles between them, so a SymmetrySite that uses both rotation and translation in fractional coords is needed.  While a fractional rotation matrix can be converted to cartesian, it seems (@tristanic?) the conversion depends on the box vectors and so it would have to be re-calculated every time the barostat changes the box vectors; a fractional rotation matrix can stay the same throughout.

Conversely, for something like a spiral with axis along z, and orthogonal box vectors a = b, c = some length*z, again with a barostat; this could be set up as either fractional or cartesian.  If fractional, then any barostat move such that a is no longer perpendicular to b, or norm(a) != norm(b), would distort the copies of the primary molecule (same as your hexamer example).  If cartesian, then it seems any barostat move that changes c would fail.  So fractional is probably more useful, it just needs a specific barostat.

So it seems there are directions along which things truly repeat via translation (primitive vectors in the Bravais lattice) - those are best handled as fractional; and there are other directions in which we may have some non-translational symmetry but we have to also repeat via translation as an artifact of the simulation setup (periodic boundary/Ewald etc) - those are best handled as cartesian to avoid deformation.  A crystal has three directions it needs to handle as fractional; a spiral has one direction where it needs fractional and the other two cartesian; a 2d lattice has two directions where it needs fractional and one cartesian; your hexamer with C6 symmetry example or eg an icosahedron has three cartesian directions, etc.  Of these fractional is more versatile: using fractional for a direction which is better described with cartesian only requires a more careful barostat that would not distort; using cartesian for a direction which is better described with fractional prevents the barostat from working.

This is just thinking out loud, I hope that helps understand the application side.

**Comment by peastman** [comment]
Possibly I'm complicating things by trying to solve two problems at once.  Consider these cases.

1. You're simulating a crystal.  The unit cell contains multiple copies of a molecule.  You want to simulate just one copy and build the others from symmetry.  In this case, it makes sense to use fractional coordinates for everything.  Is it true that this case will only ever involve rotations by 180 degrees, or are others possible?
2. You're simulating a multimer.  You want to simulate one copy and build the others from symmetry.  This doesn't necessarily have anything to do with a crystal or a unit cell.  It could even be used with implicit solvent and no periodic boundary conditions at all.

So that suggests using the boolean flag.  Either the rotation and translation are both done in Cartesian coordinates, or both done in fractional coordinates.

But what if you want to do both at once?  What if you have a multimer, and multiple copies of it fill the unit cell?  Possibly the answer is that that case is too much trouble to be worth bothering about.

**Comment by tristanic** [comment]
> Is it true that this case will only ever involve rotations by 180 degrees, or are others possible?

No. 2-, 3-, 4-, and 6-fold symmetry operations are all possible - so 180, 120, 90, and 60 degrees. 

> What if you have a multimer, and multiple copies of it fill the unit cell?

This case does occur, and can occasionally get quite extreme (e.g. 12 copies of a chain in the asymmetric unit, 6 asymmetric units making up the unit cell). Whether you'd actually want them all to be identical is very situation-dependent.

**Comment by tristanic** [comment]
In general simulations like this will be much, much easier in the NVT ensemble where the transforms are constant in both Cartesian and fractional coordinates. The moment you allow the box to stretch it all gets more difficult. As @peastman points out, if the cell dimensions change then the fractional coordinate transforms will begin to distort the molecule - which is probably why in real-world protein crystals an increase in the unit cell size almost invariably comes with a loss of resolution. The protein of course can't _actually_ expand or contract like that, so what actually drives an expansion is more solvent coming in, allowing each protein chain to rattle a bit in its lattice position, breaking the high-resolution symmetry. Conversely, a common trick to improve crystal resolution is carefully controlled dehydration - shrinking the unit cell by just a couple of Angstroms can dramatically improve things. 

For my applications, I can't currently see a situation where I _wouldn't_ be working with a fixed box size (when I use periodic boundary conditions at all - most day-to-day use will only involve rebuilding small local regions at a time). For situations where you do allow the box to change size, I'm not entirely convinced that symmetry _con_straints are the way to go.

**Comment by aizvorski** [comment]
> Either the rotation and translation are both done in Cartesian coordinates, or both done in fractional coordinates.

@peastman Yes - this sounds like a very usable API :+1: 

>  What if you have a multimer, and multiple copies of it fill the unit cell?

Hmm, would it be possible to build the multimer with cartesian, and then fill the cell with fractional?  Can a SymmetrySite depend on another SymmetrySite?

I agree with @tristanic, a lot of the more complicated cases would probably be done in NVT only so this may not matter too much.  I mainly want to make sure that it is possible to support use cases like minimizing a molecular crystal including the box vectors, or adjusting the surface tension of a 2d lattice, but it seems those would be handled well simply by having "rotation and translation both done in fractional coordinates" + the existing barostats.

**Comment by peastman** [comment]
Ok, that's what I'll implement.  Once I get a PR in you can try it out and see how well it works for you in practice.  This would be easy to change if we conclude a different behavior is needed.

**Comment by peastman** [comment]
#4185 has been merged, which means we can finally close this!
